### ISSUE TITLE
Support injection of variable determined at runtime (Node Instance ID) into userdata template (passed as --node-labels via bootstrap.sh --kubelet-extra-args)

### ISSUE THREAD

**Issue by timblaktu** [aws-ia/terraform-aws-eks-blueprints#1073]
<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

#### What is the outcome that you are trying to reach?

I want my managed nodes to be named by their EC2 instance id. To achieve this, I'm taking [this approach](https://jasonstitt.com/kubernetes-eks-tag-nodes-instance-id), and trying to set the node labels using `pre_userdata` and/or `post_userdata` in my nodegroups config to fetch the node's instance id:

```
# Early, using pre_userdata, fetch my instance id from instance metadata:
INSTANCE_ID=$(curl http://169.254.169.254/latest/meta-data/instance-id)
```

Then add `--node-labels=instance=$${INSTANCE_ID}` to the list of `--kubelet-extra-args` passed to the `bootstrap.sh` script in userdata. 

This part is the problem.

It appears this list is currently generated from the `kubelet_extra_args` terraform input variable to the `aws_eks_node_groups` contained in the `managed_ng` input to the `aws-eks-managed-node-groups` submodule, the `--kubelet-extra-args` in my actual userdata ends up looking like:

```
--kubelet-extra-args '--node-labels=eks.amazonaws.com/nodegroup-image=ami-xyz,eks.amazonaws.com/sourceLaunchTemplateVersion=3,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/nodegroup=managed-ondemand-m5large-lt-123,eks.amazonaws.com/sourceLaunchTemplateId=lt-123 --max-pods=29'
```

so I see that the terraform variable `kubelet_extra_args` is being filled by the submodule, but there's no way I can see to inject something determined at runtime, like my `--node-labels=instance=<my-dynamically-determined-instance-id` to the list of `--kubelet-extra-args` passed to the `bootstrap.sh` script in userdata.

#### Describe the solution you would like

Since [the current template for my ami_type](https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/1aac0905a12f56e2f2995ea09106efe972ab003b/modules/aws-eks-managed-node-groups/templates/userdata-amazonlinux2eks.tpl#L38) clearly substitutes the `kubelet_extra_args` terraform variable value verbatim into what gets passed  to `--kubelet-extra-args`, I thought I could just do this in my node group config:

```
      # This value is appended to the --kubelet-extra-args arg passed to bootstrap.sh.
      # Note the use of double $$ to delay variable interpolation of INSTANCE_ID until runtime, per
      #     https://developer.hashicorp.com/terraform/language/expressions/strings#escape-sequences-1
      kubelet_extra_args = "--node-labels=instance=$${INSTANCE_ID}"
```

..thinking it would insert `--node-labels=instance=${INSTANCE_ID}` literal into the `bootstrap.sh` call, resulting in something like this:

```
--kubelet-extra-args '--node-labels=instance=i-0c4c273fd4402d894,eks.amazonaws.com/nodegroup-image=ami-xyz,eks.amazonaws.com/sourceLaunchTemplateVersion=3,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/nodegroup=managed-ondemand-m5large-lt-123,eks.amazonaws.com/sourceLaunchTemplateId=lt-123 --max-pods=29'
```

Alas, what I'm getting is that my `instance=i-0c4c273fd4402d894` is omitted from `--node-labels`, and I'm guessing it's because something in the submodule is overwriting my `--node-labels` with their `--node-labels`.

#### Describe alternatives you have considered

I next tried to specify `k8s_labels` in my node group config to refer to the env var `INSTANCE_ID` (which as you'll recall I set in `pre_userdata`), like this:

```
      # INSTANCE_ID var is set in pre_userdata before bootstrap.sh is called
      k8s_labels = { instance = "$${INSTANCE_ID}" }
```

..thinking that might get appended to the `--node-labels` list by the submodule, but it does not. It errors out of course because `k8s_labels` value `"${INSTANCE_ID}"` is evaluated by terraform and validated as a label (sigh):

```
main  | Error: error updating EKS Node Group (mycluster:managed-ondemand-m5large-lt-20221021180859713400000002) config: InvalidParameterException: Invalid value: ${INSTANCE_ID} : field must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')
main  | {
main  |   RespMetadata: {
main  |     StatusCode: 400,
main  |     RequestID: "d42ea2a6-5c50-4112-9706-3bb91f367678"
main  |   },
main  |   ClusterName: "mycluster",
main  |   Message_: "Invalid value: ${INSTANCE_ID} : field must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')",
main  |   NodegroupName: "managed-ondemand-m5large-lt-20221021180859713400000002"
main  | }
main  |
main  |   with module.eks_blueprints.module.aws_eks_managed_node_groups["mng_lt"].aws_eks_node_group.managed_ng,
main  |   on .terraform/modules/eks_blueprints/modules/aws-eks-managed-node-groups/main.tf line 1, in resource "aws_eks_node_group" "managed_ng":                                                                                                                                                                                                                                   main  |    1: resource "aws_eks_node_group" "managed_ng" {
```

##### Another Approach

I suppose another obvious approach would be to insert the node ec2 instance id from terraform using a module output or data source, but I had assumed this wouldn't/couldn't work bc the instance isn't created at the time the launch template is created. Am I wrong here?

##### Yet Another (Better) Approach

Instead of reverse-engineering the way all the modules are passing and evaluating variables, **how about if we just simply added this to [the template](https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/1aac0905a12f56e2f2995ea09106efe972ab003b/modules/aws-eks-managed-node-groups/templates/userdata-amazonlinux2eks.tpl) and exposed it as a controllable feature of [modules/aws-eks-managed-node-groups](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/1aac0905a12f56e2f2995ea09106efe972ab003b/modules/aws-eks-managed-node-groups) just like `pre_userdata`, `format_mount_nvme_disk`, and `post_userdata`?**



#### Additional context

I'm using version 4.7.0 of the module and:

```
main  | Initializing provider plugins...
main  | - Reusing previous version of hashicorp/null from the dependency lock file
main  | - Reusing previous version of hashicorp/helm from the dependency lock file
main  | - Reusing previous version of hashicorp/random from the dependency lock file
main  | - Reusing previous version of hashicorp/tls from the dependency lock file
main  | - Reusing previous version of hashicorp/cloudinit from the dependency lock file
main  | - Reusing previous version of hashicorp/aws from the dependency lock file
main  | - Reusing previous version of hashicorp/time from the dependency lock file
main  | - Reusing previous version of gavinbunney/kubectl from the dependency lock file
main  | - Reusing previous version of hashicorp/kubernetes from the dependency lock file
main  | - Reusing previous version of terraform-aws-modules/http from the dependency lock file
main  | - Reusing previous version of hashicorp/local from the dependency lock file
main  | - Using previously-installed hashicorp/random v3.4.3
main  | - Using previously-installed hashicorp/tls v3.4.0
main  | - Using previously-installed hashicorp/cloudinit v2.2.0
main  | - Using previously-installed hashicorp/aws v4.34.0
main  | - Using previously-installed hashicorp/time v0.9.0
main  | - Using previously-installed hashicorp/kubernetes v2.14.0
main  | - Using previously-installed hashicorp/null v3.1.1
main  | - Using previously-installed hashicorp/helm v2.7.1
main  | - Using previously-installed gavinbunney/kubectl v1.14.0
main  | - Using previously-installed terraform-aws-modules/http v2.4.1
main  | - Using previously-installed hashicorp/local v2.2.3
```

**Comment by timblaktu** [comment]
Not stale. Have assigned this to a teammate and will provide more info ASAP.

**Comment by bryantbiggs** [comment]
this is due to the way that EKS managed node groups control and manage the bootstrap script invocation. There are workarounds available though https://github.com/awslabs/amazon-eks-ami/issues/844

**Comment by timblaktu** [comment]
Thanks @bryantbiggs for putting me on the right trail to solve that mystery. Seeing [your doc updates in the aws_eks module](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/examples/eks_managed_node_group/README.md#container-runtime--user-data), I see that:

1. It is the EKS Managed Node Group service in the default eks ami which prevents customization of the execution of the bootstrap.sh script via userdata, by overwriting at runtime (at least) the userdata section that calls bootstrap.sh.
2. Until the merging of [this PR](https://github.com/awslabs/amazon-eks-ami/pull/855/files), which adds support for sourcing an optional `bootstrap-env.sh` script and fixing `bootstrap.sh` to not overwrite `KUBELET_EXTRA_ARGS`, one must hack away at the actual `bootstrap.sh` script with sed.

I interpret this to mean that I _should_ be able to achieve what I want (customizing node labels to include instance ids) by simply specifying `pre_userdata` in my `managed_node_groups` input to eks_blueprints and that modifying eks_blueprints itself _should be_ unnecessary. 

Since I'm trying to prepend to one of the args contained by `KUBELET_EXTRA_ARGS` (`--node-labels`) instead of just appending to `KUBELET_EXTRA_ARGS` itself, I'm using a slightly different `sed` incantion, using Bash parameter expansion with regex pattern replacement. I'm not otherwise changing anything in `KUBELET_EXTRA_ARGS`, so I must also ensure that this env var is set, and I follow the pattern of sourcing a custom script:

```
      pre_userdata = <<-EOT
        #!/bin/bash
        set -ex
        # Insert "instance=$${INSTANCE_ID}," into the --node-labels arg passed to kubelet config bootsrap process in /etc/eks/bootstrap.sh.
        # See here for details on why it's necessary to overwrite bootstrap script like this:
        #     https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/examples/eks_managed_node_group/README.md#container-runtime--user-data
        #     https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/1073#issuecomment-1339537583
        #     https://github.com/awslabs/amazon-eks-ami/issues/844
        EC2_METADATA_API_TOKEN=$(curl --silent -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")
        INSTANCE_ID=$(curl -H "X-aws-ec2-metadata-token: $${EC2_METADATA_API_TOKEN}" --silent http://169.254.169.254/latest/meta-data/instance-id)
        cat <<-EOF > /etc/profile.d/bootstrap.sh
        # environment variables to be sourced in /etc/eks/bootstrap.sh
        export KUBELET_EXTRA_ARGS="$${KUBELET_EXTRA_ARGS\/--node-labels=\/--node-labels=instance=$${INSTANCE_ID},}"
        EOF
        # Patch bootstrap.sh to source extra environment variables
        sed -i '/^set -o errexit/a\\nsource /etc/profile.d/bootstrap.sh' /etc/eks/bootstrap.sh
        # Patch bootstrap.sh to respect/use existing KUBELET_EXTRA_ARGS env var
        sed -i 's/KUBELET_EXTRA_ARGS=$2/KUBELET_EXTRA_ARGS="$2 $KUBELET_EXTRA_ARGS"/' /etc/eks/bootstrap.sh
      EOT
```

Upon creation of the nodes' launch template, I confirmed in the console that the final, generated userdata looks as expected (I think):

```
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="//"

--//
Content-Type: text/x-shellscript; charset="us-ascii"
#!/bin/bash
set -ex

# User-supplied pre userdata
#!/bin/bash
set -ex
# Insert "instance=${INSTANCE_ID}," into the --node-labels arg passed to kubelet config bootsrap process in /etc/eks/bootstrap.sh.
# See here for details on why it's necessary to overwrite bootstrap script like this:
#     https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/examples/eks_managed_node_group/README.md#container-runtime--user-data
#     https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/1073#issuecomment-1339537583
#     https://github.com/awslabs/amazon-eks-ami/issues/844
EC2_METADATA_API_TOKEN=$(curl --silent -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")
INSTANCE_ID=$(curl -H "X-aws-ec2-metadata-token: ${EC2_METADATA_API_TOKEN}" --silent http://169.254.169.254/latest/meta-data/instance-id)
cat <<-EOF > /etc/profile.d/bootstrap.sh
# environment variables to be sourced in /etc/eks/bootstrap.sh
export KUBELET_EXTRA_ARGS="${KUBELET_EXTRA_ARGS\/--node-labels=\/--node-labels=instance=${INSTANCE_ID},}"
EOF
# Patch bootstrap.sh to source extra environment variables
sed -i '/^set -o errexit/a\\nsource /etc/profile.d/bootstrap.sh' /etc/eks/bootstrap.sh
# Patch bootstrap.sh to respect/use existing KUBELET_EXTRA_ARGS env var
sed -i 's/KUBELET_EXTRA_ARGS=$2/KUBELET_EXTRA_ARGS="$2 $KUBELET_EXTRA_ARGS"/' /etc/eks/bootstrap.sh

--//--
```

However I'm not seeing my "patches" in the `/etc/eks/bootstrap.sh` in newly-created node instances. Any ideas what I did wrong here?

```

**Comment by timblaktu** [comment]
For future onlookers frustrated by 3 levels of nested character escape sequence rules (terraform, bash, sed), this is what I ended up on that works, properly hacking the setting of KUBELET_EXTRA_ARGS inside bootstrap.sh:

```
      pre_userdata = <<-EOT
        #!/bin/bash
        #
        # Patch bootstrap.sh handling of --kubelet-extra-args to prepend our desired label to the --node-labels embedded in it.
        # This script hacking is required because the EKS managed node group service in the default AMI is overwriting parts of the userdata
        # most notably the line that calls `/etc/eks/bootstrap.sh`, in order to inject --kubelet-extra-args generated from external configuration:
        # --node-labels, --max-pods, --b64-cluster-ca, etc. Also, because `bootstrap.sh` doesn't currently respect the KUBELET_EXTRA_ARGS env var,
        # the only option is to hack the line _inside_ `bootstrap.sh` which handles the `--kubelet-extra-args` arg, so that it prepends
        # 'instance=$INSTANCE_ID,' into the comma-separated `--node-labels` list passed into `bootstrap.sh` inside the `--kubelet-extra-args` arg.
        # See here for details on why it's necessary to overwrite bootstrap script like this:
        #     https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/examples/eks_managed_node_group/README.md#container-runtime--user-data
        #     https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/1073#issuecomment-1339537583
        #     https://github.com/awslabs/amazon-eks-ami/issues/844
        set -ex
        EC2_METADATA_API_TOKEN=$(curl --silent -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")
        INSTANCE_ID=$(curl -H "X-aws-ec2-metadata-token: $${EC2_METADATA_API_TOKEN}" --silent http://169.254.169.254/latest/meta-data/instance-id)
        sed -i "s/KUBELET_EXTRA_ARGS=\$2/KUBELET_EXTRA_ARGS=\"\$${2\/--node-labels=\/--node-labels=instance=$INSTANCE_ID,}\"/" /etc/eks/bootstrap.sh
      EOT
```

Even more time consuming for me here was the fact that I wrongly thought I needed to set `create_launch_template = true` for the `pre_userdata` to work, and was (and still am) getting errors with nodes not connecting to the control plane. Once i set `create_launch_template = false` the issues went away. I _would_ however like to get the launch template working, but of course that's a separate issue...

**Comment by timblaktu** [comment]
Reopening bc although I've got `pre_userdata` script functionally doing what I need, now I can't get my managed nodes to join the cluster as long as I have `create_launch_template = true`. @bryantbiggs @Zvikan could you take a quick look at my blueprints config at the bottom and let me know if you see any configuration issue? 

I can `aws ssm start-session` into the nodes before the group creation times out, and see that their userdata and my bootstrap.sh hack look good, but in the kubelet logs I see this:

```
Dec 15 22:35:21 ip-10-0-11-202.us-west-1.compute.internal kubelet[3165]: E1215 22:35:21.023225    3165 kubelet.go:2469] "Error getting node" err="node \"ip-10-0-11-202.us-west-1.compute.internal\" not found"
Dec 15 22:35:21 ip-10-0-11-202.us-west-1.compute.internal kubelet[3165]: E1215 22:35:21.083298    3165 kubelet.go:2394] "Container runtime network not ready" networkReady="NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized"
```

It appears this is caused by the CNI not being configured in the cluster. I believe this to be something blueprints took care of when I was using `create_launch_template = false`, and I must be missing some essential bit of config now that I've set it `true`.

It looks like the blueprints `vpc-cni` submodule is only created when the [enable_amazon_eks_vpc_cni input variable to modules/kubernetes-addons](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/c309d3bf7588ad43c3319ac320013609715e7574/modules/kubernetes-addons#input_enable_amazon_eks_vpc_cni) is set. Since I'm not setting this, and it defaults to false, I’m assuming we haven’t needed to explicitly use this module since something else in blueprints is doing that. I first assumed it was probably [the aws_eks module underneath](https://github.com/terraform-aws-modules/terraform-aws-eks#usage), which has [input variable cluster_addons](https://github.com/terraform-aws-modules/terraform-aws-eks#input_cluster_addons) that can be used to specify the vpc-cni addon to be installed. However, this also defaults to off/empty, and [here you can see that eks_blueprints does not even specify cluster_addons when invoking the eks module](https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/90346b8545072ac032f64c232cbcaa6950348034/main.tf#L21), so I’m perplexed ATM about how/if the vpc-cni is being created.

Here's my config. It's mostly copied from the mng lt example. Note it's using my fork, but this is basically blueprints 4.15 with minor changes made to crossplane module to add unrelated support for helm/terraform providers:

```
module "eks_blueprints" {
  # source = "github.com/aws-ia/terraform-aws-eks-blueprints?ref=v4.15.0"
  source = "github.com/timblaktu/terraform-aws-eks-blueprints?ref=crossplane-helm-provider"
  cluster_name    = local.name
  cluster_version = local.cluster_version
  vpc_id             = module.vpc.vpc_id
  private_subnet_ids = module.vpc.private_subnets
  cluster_endpoint_private_access = true
  cluster_endpoint_public_access = true
  create_cluster_security_group = false
  cluster_security_group_id     = aws_security_group.eks.id
  node_security_group_additional_rules = {
    ingress_self_all = {
      description = "Node to node all ports/protocols"
      protocol    = "-1"
      from_port   = 0
      to_port     = 0
      type        = "ingress"
      self        = true
    }
    egress_all = {
      description      = "Node all egress"
      protocol         = "-1"
      from_port        = 0
      to_port          = 0
      type             = "egress"
      cidr_blocks      = ["0.0.0.0/0"]
      ipv6_cidr_blocks = ["::/0"]
    }
    ingress_cluster_to_node_all_traffic = {
      description                   = "Cluster API to Nodegroup all traffic"
      protocol                      = "-1"
      from_port                     = 0
      to_port                       = 0
      type                          = "ingress"
      source_cluster_security_group = true
    }
  }
  cluster_kms_key_arn = data.aws_kms_key.eks_secrets.arn

  managed_node_groups = {
    mng_lt = {
      node_group_name = "managed-ondemand-m5large-lt"
      ami_type               = "AL2_x86_64"
      release_version        = ""
      create_launch_template = true
      launch_template_os     = "amazonlinux2eks"
      capacity_type          = "ON_DEMAND"  
      instance_types         = ["m5.large"]
      format_mount_nvme_disk = true       
      desired_size           = 2
      max_size               = 2
      min_size               = 2

      create_iam_role         = false
      iam_role_arn            = aws_iam_role.managed_ng.arn
      additional_iam_policies = []

      subnet_type = "private"
      subnet_ids  = module.vpc.private_subnets
      public_ip   = false

      enable_monitoring = true
      eni_delete        = true

      http_endpoint               = "enabled"
      http_tokens                 = "optional"
      http_put_response_hop_limit = 3

      update_config = [{
        max_unavailable_percentage = 30
      }]
      timeouts = [{
        create = "10m"
        update = "20m"
        delete = "20m"
      }]

      k8s_taints = []

      pre_userdata = <<-EOT
        EC2_METADATA_API_TOKEN=$(curl --silent -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")
        INSTANCE_ID=$(curl -H "X-aws-ec2-metadata-token: $${EC2_METADATA_API_TOKEN}" --silent http://169.254.169.254/latest/meta-data/instance-id)
        sed -i "s/KUBELET_EXTRA_ARGS=\$2/KUBELET_EXTRA_ARGS=\"\$${2\/--node-labels=\/--node-labels=instance=$INSTANCE_ID,}\"/" /etc/eks/bootstrap.sh
      EOT

      block_device_mappings = [
        {
          device_name = "/dev/xvda"
          volume_type = "gp3"
          volume_size = 100
        }
      ]

      additional_tags = {
        subnet_type = "private"
      }

      remote_access = false
    }
  }
  tags = local.tags
}

module "eks_blueprints_kubernetes_addons" {
  # source = "github.com/aws-ia/terraform-aws-eks-blueprints//modules/kubernetes-addons?ref=v4.15.0"
  source = "github.com/timblaktu/terraform-aws-eks-blueprints//modules/kubernetes-addons?ref=crossplane-helm-provider"

  eks_cluster_id       = module.eks_blueprints.eks_cluster_id
  eks_cluster_endpoint = module.eks_blueprints.eks_cluster_endpoint
  eks_oidc_provider    = module.eks_blueprints.oidc_provider
  eks_cluster_version  = module.eks_blueprints.eks_cluster_version

  tags = local.tags

  # enable_metrics_server     = true
  enable_cluster_autoscaler = true
  enable_argocd = true

  argocd_helm_config = {
    namespace        = "argocd"
    timeout          = "1200"
    create_namespace = true
    values           = [templatefile("${path.module}/templates/argocd-values.yaml", {})]
    set_sensitive = [
      {
        name  = "configs.secret.argocdServerAdminPassword"
        value = bcrypt_hash.argo.id
      }
    ]
  }
  argocd_manage_add_ons = true
  crossplane_aws_provider = {
    enable               = true
    provider_aws_version = "v0.33.0"
    additional_irsa_policies = [
      "arn:aws:iam::aws:policy/AdministratorAccess",
    ]
  }

  crossplane_kubernetes_provider = {
    enable                      = true
    provider_kubernetes_version = "v0.5.0"
  }
  crossplane_helm_provider = {
    enable                = true
    provider_helm_version = "v0.12.0"
  }

  crossplane_terraform_provider = {
    enable                     = true
    provider_terraform_version = "v0.5.0"
    additional_irsa_policies = [
      "arn:aws:iam::aws:policy/AdministratorAccess",
    ]
  }

  argocd_applications = var.cleanup_argocd_applications ? {} : {
    addons = {
      path               = "chart"
      repo_url           = "https://github.com/aws-samples/eks-blueprints-add-ons.git"
      add_on_application = true
    }
  }
}
```

**Comment by timblaktu** [comment]
I'm now wondering if this could be caused by [the deprecation of the docker runtime in eks ami versions past 1.23](https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html). Since I'm using: 
* `cluster_version        = 1.23`
* `ami_type               = "AL2_x86_64"`
* `release_version        = ""`
* `create_launch_template = true`
* `launch_template_os     = "amazonlinux2eks"`

..and *NOT* specifying to use containerd anywhere, maybe I'm getting a 1.24 ami, which doesn't support docker, and my kubelets are trying to use docker runtime.

If this were true, it would be a bug IMO since the ami `release_version` should be selected to correspond to the `cluster_version` in cases where you're not specifying a `custom_ami_id`. Thoughts, @bryantbiggs @Zvikan?

**Comment by timblaktu** [comment]
Next area to double-check was public/private endpoint access and associated DNS settings in my VPC. Since I had recently added these two lines in my `eks_blueprints` config:

```
  cluster_endpoint_private_access = true
  cluster_endpoint_public_access = true
```

and the previous defaults being used were, effectively:

```
  cluster_endpoint_private_access = false
  cluster_endpoint_public_access = true
```

..I had effectively turned on private endpoint access to restrict all node<-->controlplane traffic to intra-vpc.

Reverting back to the default settings (only public endpoint enabled), I saw some improvement in node behavior. On node group creation, instead of not ever seeing the nodes populated in the nodegroup console (and having to look their instance ids up using the cluster name in ec2 console), I now am seeing the nodes immediately, albeit they are still not joining due to same node condition error reported above:

```
Container runtime network not ready" networkReady="NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized"
```

I'm still flummoxed as to why the `vpc-cni` addon isn't being installed by default, or in any of the examples (aside from the "custom networking" or "custom ami" types which I'm not using), and what are the circumstances under which it needs to be explicitly added to or enabled in eks_blueprints config. 

So, I decided to just add this to my eks_blueprints_kubernetes_addons config:

```
  enable_amazon_eks_vpc_cni = true
  amazon_eks_vpc_cni_config = {
    most_recent       = true
    resolve_conflicts = "OVERWRITE"
  }
```

and *_for the first time since setting `create_launch_template = true`, my nodes successfully joined the cluster_* (_as well as setting my node labels via custom userdata per this original issue_)! So, my questions are:

1. Why/when is it necessary to explicitly `enable_amazon_eks_vpc_cni`?
2. Is it necessary to explicitly `enable_amazon_eks_vpc_cni` when using a node group launch template with an `ami_type` but no `custom_ami_id`?
3. When not using a node group launch template, and letting `enable_amazon_eks_vpc_cni` default to `false`, why did I not get this same "cni config uninitialized" error?
